sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala

```diff
@@ -855,4 +874,30 @@ abstract class BucketedReadSuite extends QueryTest with SQLTestUtils {
       }
     }
   }
+
+  test("bucket coalescing eliminates shuffle") {
```

Comment:
We need to add more exhaustive tests for this optimization by using `COALESCE_BUCKET_IN_JOIN_MAX_NUM_BUCKETS_DIFF `, e.g., boundary config value tests.

Reply:
Still working on this, will update tests in the next iteration.

Reply:
Added more tests.